components/feature/search/src/main/java/mozilla/components/feature/search/ext/SearchEngine.kt

```diff
@@ -66,3 +72,49 @@ fun parseLegacySearchEngine(id: String, stream: InputStream): SearchEngine {
     val reader = SearchEngineReader(SearchEngine.Type.CUSTOM)
     return reader.loadStream(id, stream)
 }
+
+/**
+ * Given a [SearchState], determine if the passed-in [url] is a known search results page url
+ * and what are the associated search terms.
+ * @return Search terms if [url] is a known search results page, `null` otherwise.
+ */
+fun SearchState.parseSearchTerms(url: String): String? {
+    val parsedUrl = Uri.parse(url)
+    // Default/selected engine is the most likely to match, check it first.
+    val currentEngine = this.selectedOrDefaultSearchEngine
+    // Or go through the rest of known engines.
+    val fallback: () -> String? = fallback@{
+        this.searchEngines.forEach { searchEngine ->
+            searchEngine.parseSearchTerms(parsedUrl)?.let { return@fallback it }
+        }
+        return@fallback null
+    }
+    return currentEngine?.parseSearchTerms(parsedUrl) ?: fallback()
+}
```

Comment:
Note that this will not work for every search engine. Some (like Wikipedia) do a redirect and the resulting URL will not match the build URL anymore.